arch/xtensa/core/xtensa_asm2_util.S

```diff
@@ -83,7 +83,7 @@ _high_gpr_spill_done:
 	 * time how many registers were spilled, then return, leaving the
 	 * modified SP in A1.
 	 */
-	addi a1, a1, -4
+	addi a1, a1, -16
```

Comment:
This doesn't do anything to align SP, just offsets it differently.  The interrupt can have occurred with the stack at any state, at least in theory.  If you want to align it you want to mask off the bottom bits after the subtraction, not just change the offset.

Reply:
@andyross Based on Xtensa ABI requirements stack needs to be 16 bytes aligned. I see that xtensa compiler actually require it.

Let's take simple example:
```
int main() {
    __attribute__((aligned(16))) volatile  int a = 1;
    printf("%p = %d\n", &a, a);
    return 0;
}
```

I am getting disassembly:
```
20000ec8 <main>:
20000ec8:       008136                                          entry   a1, 64
20000ecb:       020c                                            movi.n  a2, 0
20000ecd:       7129                                            s32i.n  a2, a1, 28
20000ecf:       0020c0                                          memw
20000ed2:       120c                                            movi.n  a2, 1
20000ed4:       4129                                            s32i.n  a2, a1, 16
20000ed6:       0020c0                                          memw
20000ed9:       0421c2                                          l32i    a12, a1, 16
20000edc:       2000a4                                          const16 a10, 0x2000
20000edf:       0bd0a4                                          const16 a10, 0xbd0
20000ee2:       10c1b2                                          addi    a11, a1, 16
20000ee5:       000065                                          call8   20000eec <printf>
20000ee8:       020c                                            movi.n  a2, 0
20000eea:       f01d                                            retw.n
```

If `a1` at entry is not 16-bytes aligned it won't work.

NOTE 1: Alignments over 16 bytes are handled differently, but it will only work with 16 bytes alignment.
NOTE 2: every entry a1 will reserve space on stack with 16-bytes increments: (entry a1, 16/32/48 etc)

Also, `addi a1, a1, -4` breaks alignment. Before executing this instruction stack is 16-bytes aligned, post this instruction it is not.

Reply:
will contact cadence to have that requirement in writing

Reply:
Hm... I tried it on both `samples/hello_world` (with `qemu_xtensa`) and `samples/userspace/hello_world_user` (with `qemu_xtensa/dc233c/mmu`). They both worked without issue.

Reply:
> Also, addi a1, a1, -4 breaks alignment. Before executing this instruction stack is 16-bytes aligned

No, it's aligned however it was when you entered the trap.  Xtensa takes its interrupts on the interrupted context's stack, and the interrupted context may have done all sorts of things to A1.    If you want the stack aligned you need to actually align the value (e.g. `A1 := (A1 - 4) & ~7`)

Reply:
But to be clear, I buy the argument that (even if Cadence themselves aren't super clear) the toolchain is assuming 16 byte alignment, clearly the generated assembly isn't doing any alignment math.

@dcpleung the reason it worked is presumably that you added the code to main(), which is running in thread mode and aligned correctly; if you put it in an ISR (like a k_timer callback), I'd guess you'd see the problem.

Reply:
Hm... maybe we are just lucky that `arch_new_thread()` results in correct alignment under QEMU. There is no explicit alignment being done there.

Reply:
> > Also, addi a1, a1, -4 breaks alignment. Before executing this instruction stack is 16-bytes aligned
> 
> No, it's aligned however it was when you entered the trap. Xtensa takes its interrupts on the interrupted context's stack, and the interrupted context may have done all sorts of things to A1. If you want the stack aligned you need to actually align the value (e.g. `A1 := (A1 - 4) & ~7`)

if user intentionally modifies A1 to be invalid, I wouldn't recommend fixing it. Compiler by default will align it to 16-bytes boundaries.
Only way to break it is to use assembly or maybe `register` keyword in C/C++

Reply:
App code has access to `asm()` last I checked.  Come on, if this is worth doing it's worth doing right.